cmd/registry/scoring/score.go

```diff
@@ -0,0 +1,366 @@
+	definition *rpc.ScoreDefinition,
+	resource patterns.ResourceInstance) (interface{}, error) {
+
+	// Apply score formula
+	switch formula := definition.GetFormula().(type) {
+	case *rpc.ScoreDefinition_ScoreFormula:
+		scoreValue, err := processScoreFormula(ctx, client, formula.ScoreFormula, resource)
+		if err != nil {
+			return nil, err
+		}
+		return scoreValue, nil
+	// TODO: Add support for rollup_formula
+	default:
+		return nil, fmt.Errorf("invalid formula in ScoreDefinition: {%v} ", formula)
+	}
+}
+
+func processScoreFormula(
+	ctx context.Context,
+	client connection.Client,
+	formula *rpc.ScoreFormula,
+	resource patterns.ResourceInstance) (interface{}, error) {
+
+	// Fetch the artifact
+	extendedArtifact, err := patterns.SubstituteReferenceEntity(formula.GetArtifact().GetPattern(), resource.ResourceName())
+	if err != nil {
+		return nil, fmt.Errorf("invalid score_formula.artifact.pattern: %s, %s", formula.GetArtifact().GetPattern(), err)
+	}
+	artifactName, err := names.ParseArtifact(extendedArtifact.String())
+	if err != nil {
+		return nil, fmt.Errorf("invalid score_formula.artifact.pattern: %s, %s", formula.GetArtifact().GetPattern(), err)
+	}
+
+	var contents []byte
+	var mimeType string
+	err = core.GetArtifact(ctx, client, artifactName, true, func(artifact *rpc.Artifact) error {
+		contents = artifact.GetContents()
+		mimeType = artifact.GetMimeType()
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch artifact %s: %s", artifactName, err)
+	}
+
+	// TODO: Add timestamp check. Compute the score only if the artifact has an update since the last score calculation
+
+	// Apply the score_expression
+	value, err := evaluateScoreExpression(formula.GetScoreExpression(), mimeType, contents)
+	if err != nil {
+		return nil, err
+	}
+
+	return value, nil
+}
+
+func processScoreType(definition *rpc.ScoreDefinition, scoreValue interface{}, project string) (*rpc.Score, error) {
+
+	// Initialize Score proto
+	score := &rpc.Score{
+		Id:             fmt.Sprintf("score-%s", definition.GetId()),
+		Kind:           "Score",
+		DisplayName:    definition.GetDisplayName(),
+		Description:    definition.GetDescription(),
+		Uri:            definition.GetUri(),
+		UriDisplayName: definition.GetUriDisplayName(),
+		DefinitionName: fmt.Sprintf("%s/artifacts/%s", project, definition.GetId()),
+	}
+
+	// Set the Value field according to the type
+	switch definition.GetType().(type) {
+	case *rpc.ScoreDefinition_Integer:
+		// Score proto expects int32 type
+		var value int32
+
+		// Convert scoreValue to appropriate type
+		// evaluateScoreExpression can return either a float or int value.
+		// Both are valid for an integer.
+		if intVal, ok := scoreValue.(int64); ok {
+			value = int32(intVal)
+		} else if floatVal, ok := scoreValue.(float64); ok {
+			value = int32(floatVal)
+		} else {
+			return nil, fmt.Errorf("failed typecheck for output: expected either int or float64 got %s", scoreValue)
+		}
```

Comment:
I think a type switch is easier to visually parse for something like this.

```
switch v := scoreValue.(type) {
case int64, float64:
	value = int32(v)
default:
	return nil, fmt.Errorf("failed typecheck for output: expected either int64 or float64 got %T %s", v, v)
}
```

Also, I think the error message can be more specific about the types it expects, and show what type it received (rather than just the value).

Same for percentages below.

Reply:
This switch case block fails at typecasting `int32(v)`. I think this is happening because the type of `v` is known only at runtime and this code fails to build because at build time, the type of `v` is still `interface{}`. Even in the `switch` block, I will need a typecheck, which will bring it close to the current `if else` structure. Example: https://go.dev/play/p/zWNXMHsJuON

Updated the error message as suggested.

Reply:
I suppose that's what I get for writing that out without running it. Type switches in Go automatically typecast the switch variable (if one is defined, `v` in this case) to the type in each switch case. My mistake was thinking it would know that `int32(v)` works in both cases.

Splitting it into separate cases should work as intended.
```
switch v := scoreValue.(type) {
case int64:
	value = int32(v)
case float64:
	value = int32(v)
default:
	return nil, fmt.Errorf("failed typecheck for output: expected either int64 or float64 got %T %s", v, v)
}
```

Reply:
Thanks for clarifying. It didn't strike me to try for different cases. 
Works as expected!